From f4ff4aed12d12e496488d01ab821fc6ac2002229 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Tue, 8 Oct 2024 15:40:08 +0200
Subject: [PATCH 1337/1695] drivers: stm32_rng: add RNG support for STM32MP21
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

MP21 needs specific NSCR, HTCR and CR values to work properly.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 366792164abb744a04ccbf5720d9556f0838e2e7 drivers: stm32_rng: add RNG support for STM32MP21]
Signed-off-by: Maxime Méré <maxime.mere@foss.st.com>
Change-Id: I0438526e6fefd52f91a1eb97cee97ae9e5bb0336
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/409820
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Tested-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/stm32_rng.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/core/drivers/stm32_rng.c b/core/drivers/stm32_rng.c
index 468e810f2..b33a593fd 100644
--- a/core/drivers/stm32_rng.c
+++ b/core/drivers/stm32_rng.c
@@ -732,6 +732,18 @@ static const struct stm32_rng_driver_data mp15_data[] = {
 };
 DECLARE_KEEP_PAGER_PM(mp15_data);
 
+static const struct stm32_rng_driver_data mp21_data[] = {
+	{
+		.max_noise_clk_freq = U(48000000),
+		.nb_clock = 2,
+		.has_cond_reset = true,
+		.has_power_optim = true,
+		.cr = 0x00800D00,
+		.nscr = 0x01F7,
+		.htcr = 0xAAC7,
+	},
+};
+
 static const struct stm32_rng_driver_data mp25_data[] = {
 	{
 		.max_noise_clk_freq = U(48000000),
@@ -747,6 +759,7 @@ static const struct stm32_rng_driver_data mp25_data[] = {
 static const struct dt_device_match rng_match_table[] = {
 	{ .compatible = "st,stm32-rng", .compat_data = &mp15_data },
 	{ .compatible = "st,stm32mp13-rng", .compat_data = &mp13_data },
+	{ .compatible = "st,stm32mp21-rng", .compat_data = &mp21_data },
 	{ .compatible = "st,stm32mp25-rng", .compat_data = &mp25_data },
 	{ }
 };
-- 
2.47.2

